docs/plutus-report/scripting.tex

```diff
@@ -111,7 +115,127 @@ \subsection{Recursive values}
+
+\paragraph{Polymorphic builtin types}
+
+It would be nice to support polymorphic builtin types.
+For example, that would allow us to provide optimized implementations of container types such as maps.
+
+This seems feasible in principle, but we have not worked the details out yet.
+
+\subsubsection{Builtin functions}
+
+We could implement extensible builtin functions in a similar way to extensible builtin types: use some universe that catalogues the available builtin functions.
+However, this is increasingly tedious, and we don't have the same dependency on the available functions as we do on the types, since we don't have anything corresponding to constants.
+
+So instead we adopt a much simpler approach where the available builtin functions can be provided at runtime simply indexed by a name.
+
+\paragraph{Passing arguments to builtin functions}
+
+Builtin functions must of course be provided with a type signature (since we have a typed language), and a way to evaluate them.
+The natural way to provide such evaluators is as functions --- but these functions must be able to operate on \gls{plutus-core} terms!
+For example, an evaluator for ``plus'' must be able to take two \emph{terms} of type \code{Integer} and turn them into another term.
+But a simple implementation of ``plus'' just talks about \emph{integers}, not terms.
+
+Solving this problem in generality is difficult, so we make a key simplification: builtin functions may only accept arguments of builtin types (with one exception, as we will see shortly).
+Along with the fact that (since \gls{plutus-core} is eager) function arguments are evaluated to values before being passed to the function, this means that the arguments to builtin evaluators will always be \emph{constants}, which are easy to unwrap into their underlying values.
+
+\paragraph{Polymorphic builtin functions}
+\label{sec:polymorphic-builtins}
+
+The exception to the above rule about builtin function types is for \emph{polymorphic} builtin functions.
+These must necessarily be able to process arbitrary terms (since they can be instantiated at any type, not just a builtin type).
+For example, \code{ifThenElse} must be able to produce a value of type \code{Integer -> Integer}, even though this is not a builtin type.
+
+It is possible in principle to use polymorphic \emph{evaluators} to implement polymorphic builtin functions \autocite{lindley2012embedding}.
+Unfortunately this is quite a heavyweight and complex encoding for such a key usecase.
+
+Instead, we require evaluators for polymorphic functions to be \emph{parametric}, which means in particular that they do not inspect their polymorphic arguments.
+We can thus pass the \emph{term itself} as the polymorphic argument, instead of needing to extract a value of an appropriate type in the metalanguage, as we do for other arguments.
+Thus, an evaluator for \code{const :: forall a b . a -> b -> a} takes two \emph{terms} and returns the first one.
```

Comment:
"an evaluator for `<a particular term>`" reads weirdly to me.

Reply:
hm okay